File: protocol/src/ptt-semantic-state.ts
export type PttSemanticState =
  | "READY" | "TALKING" | "BUSY_REMOTE" | "PREPARING"
  | "ERROR" | "RECOVERING" | "OFFLINE" | "RX_ONLY";

export function pttSemanticState(input: {
  listener: boolean;
  connected: boolean;
  recovering: boolean;
  remoteTalking: boolean;
  pttState: "IDLE" | "REQUESTING" | "TRANSMITTING" | "BUSY" | "RELEASING";
  error: boolean;
}): PttSemanticState {
  if (input.listener) return "RX_ONLY";
  if (input.error) return "ERROR";
  if (!input.connected) return "OFFLINE";
  if (input.recovering) return "RECOVERING";
  if (input.remoteTalking || input.pttState === "BUSY") return "BUSY_REMOTE";
  if (input.pttState === "TRANSMITTING") return "TALKING";
  if (input.pttState === "REQUESTING" || input.pttState === "RELEASING") return "PREPARING";
  return "READY";
}

File: android/app/src/main/java/com/dennomuso/koeon/core/ptt/PttSemanticState.kt
package com.dennomuso.koeon.core.ptt

enum class PttSemanticState {
    READY, TALKING, BUSY_REMOTE, PREPARING, ERROR, RECOVERING, OFFLINE, RX_ONLY
}

fun localPttEligible(
    operationallyActive: Boolean,
    canPublish: Boolean,
    connected: Boolean,
    audioReady: Boolean,
    remoteTalking: Boolean,
): Boolean = operationallyActive && canPublish && connected && audioReady && !remoteTalking

fun pttSemanticState(
    canPublish: Boolean,
    connected: Boolean,
    recovering: Boolean,
    remoteTalking: Boolean,
    pttState: PttState,
): PttSemanticState = when {
    !canPublish -> PttSemanticState.RX_ONLY
    pttState == PttState.ERROR -> PttSemanticState.ERROR
    !connected -> PttSemanticState.OFFLINE
    recovering -> PttSemanticState.RECOVERING
    pttState == PttState.TRANSMITTING -> PttSemanticState.TALKING
    pttState == PttState.REQUESTING_FLOOR || pttState == PttState.RELEASING -> PttSemanticState.PREPARING
    remoteTalking || pttState == PttState.BUSY -> PttSemanticState.BUSY_REMOTE
    else -> PttSemanticState.READY
}
